content/blog/2025-05-13-flat-bril/index.md

```diff
@@ -0,0 +1,170 @@
+
+Our final representation of a function that worked with zerocopy:
+
+```Rust
+#[repr(packed)]
+#[derive(Debug, PartialEq, Clone, Immutable, IntoBytes)]
+pub struct FunctionView<'a> {
+    pub func_name: &'a [u8],
+    pub func_args: &'a [FlatFuncArg],
+    pub func_ret_ty: FlatType,
+    pub var_store: &'a [u8],
+    pub arg_idxes_store: &'a [I32Pair],
+    pub labels_idxes_store: &'a [I32Pair],
+    pub labels_store: &'a [u8],
+    pub funcs_store: &'a [u8],
+    pub instrs: &'a [FlatInstr],
+}
+```
+
+Once we finally made zerocopy happy about all our data structures we were able to convert seamlessly between the in memory representations and bytes which we could directly mmap to disk, enabling the flat file format we envisioned as one of our goals.
+
+When deserializing from our flat file format back to our flat data structures we ran into a very nasty bug. On some programs we would get alignment errors when attempting to convert bytes to our data structures, but we could fix/break programs just by changing the variable names! We eventually realized that a program would crash if the total bytes used to represent all variable names (or labels, or functions) was not a multiple of 4! This was because our top-level arrays just stored all names contiguously, and so would be unaligned unless the total bytes to represent all names was a multiple of 4. To fix this we just padded all of these top level arrays with null bytes.
+
+### Interpreting flat Bril
+
+Once we had finished implementing our infrastructure for flat Bril representations, we implemented an interpreter that operated directly on the flat data structures. We built this interpreter from the ground up rather than trying to adapt the existing Rust interpreter. This was mostly straightforward, and we were able to model our logic after the Typescript interpreter. The main challenges were introduced by needing to keep track of which array each pair of indices in our flat data structures was referring to. We had a few bugs caused by assuming that some indices referred to the top-level name arrays, when in fact we needed to go through an intermediate array. In hindsight we could have done a better job with our naming conventions to avoid this.
```

Comment:
[idle comment; no actionable change] Honestly, in my experience, this problem where you mix up the association between indices and arrays (and therefore use an index to access some array where that index is meaningless) is one of the bigger frustrations I've had when programming in this style. A little type safety could help a lot here…